test/hotspot/jtreg/serviceability/jvmti/HeapMonitor/MyPackage/HeapMonitor.java

```diff
@@ -164,7 +164,21 @@ public static void freeStorage() {
     enableSamplingEvents();
     setSamplingInterval(0);
 
-    // Loop around an allocation loop and wait until the tlabs have settled.
+    // Loop around an allocation loop and wait until the TLABs have settled.
+    // It takes two steps to enable Object Sampling with interval 0
+    // 1. consume current TLAB, which can be varies with heap/GC configuration
+    //    then allocation enters allocate_inside_tlab_slow and check
+    //    JvmtiExport::should_post_sampled_object_alloc
+    // 2. consume initial ThreadHeapSampler::_bytes_until_sample
+    //    then runtime invokes pick_next_sample and check sampleing interval
+    //    and set _bytes_until_sample to 0.
+    //    initial _bytes_until_sample is geometric variable with the specified mean
+    //    (512K by default), check ThreadHeapSampler::pick_next_geometric_sample()
```

Comment:
I think there is too much detail about the hotspot implementation in these comments. I think for (1) it is enough to say that the current TLAB needs to be consumed, and that this can be accomplished with a System.gc(). For (2) I don't really understand what you are saying. If you can explaining without referring to specific fields in hotspot, that would be better. I also don't understand how you are making sure that (2) is triggered. Does that happen during the allocation loop?

Reply:
Thanks for your patient review！New comment is simplified.

Step1 is retiring current TLAB. 

When allocating new TLAB, runtime checks if JVMTI object sample is enabled and report event.  After report runtime also calculates how many bytes to skip before perform next object sample, which is stored in ThreadHeapSampler's field _bytes_until_sample. Also need consume these bytes before runtime picks up user specified object sample interval bytes (0 in this case).
 
Step2 is an allocation loop consumes current ThreadHeapSampler's _bytes_until_sample bytes (initial value is randome size about 512K).